### PR TITLE
[data] Support block metadata and lazy loading of blocks from the filesystem

### PR DESCRIPTION
**Pull request by ericl**
This adds BlockMetadata, which is stored separately from the block reference. This allows for metadata operations like .count(), size_bytes(), .input_files(), and .schema() to avoid scanning the full data.

### CODE REVIEW

File: python/ray/experimental/data/dataset.py
Comment (scv119) [inline python/ray/experimental/data/dataset.py:297]: does this change the time complexity of limit?

Comment (scv119) [inline python/ray/experimental/data/dataset.py:297]: does this change the time complexity of limit?

File: python/ray/experimental/data/dataset.py
Comment (scv119) [inline python/ray/experimental/data/dataset.py:391]: maybe we need a todo task to validate that all the blocks has same schema to avoid surprises

**Comment by scv119** [comment]
first round of review...

File: python/ray/experimental/data/dataset.py
Comment (scv119) [inline python/ray/experimental/data/dataset.py:547]: not matching the batch_size/format yet.

File: python/ray/experimental/data/impl/block_list.py
Comment (clarkzinzow) [inline python/ray/experimental/data/impl/block_list.py:9]: Isn't `BlockMetadata` more of a per-block thing, i.e. shouldn't this be in `impl/block.py`? Especially since it'll be useful outside of just `BlockList` (I think).

File: python/ray/experimental/data/impl/compute.py
Comment (clarkzinzow) [inline python/ray/experimental/data/impl/compute.py:41]: Nit: Use `zip()` idiom.
```suggestion
        new_blocks, new_metadata = zip(*refs)
```

File: python/ray/experimental/data/impl/compute.py
Comment (clarkzinzow) [inline python/ray/experimental/data/impl/compute.py:77]: Nit:
```suggestion
        blocks_in = list(zip(blocks, blocks.get_metadata()))
```

File: python/ray/experimental/data/impl/shuffle.py
Comment (clarkzinzow) [inline python/ray/experimental/data/impl/shuffle.py:64]: Nit: Use `zip()` idiom.
```suggestion
    new_blocks, new_metadata = zip(*shuffle_reduce_out)
```

File: python/ray/experimental/data/read_api.py
Comment (clarkzinzow) [inline python/ray/experimental/data/read_api.py:23]: Nit: Should use `functools.wraps` for at least the `.__name__` and `.__doc__` attrs. https://docs.python.org/3/library/functools.html#functools.wraps

**Review by clarkzinzow — COMMENTED** [review]
[Transferred from other PR] A few comments upon a quick skim.

At a higher level and more future looking, I think that we might benefit from changing the block future abstraction from `ObjectRef[Block]` with a `BlockMetadata` sidecar object to `Block[Union[Data, ObjectRef[Data]], Metadata]`, where the block wraps a data (future) object and block metadata that's known at `Block` instantiation time, and exposes pretty much the same APIs. I think that something like that would simplify a lot of the remaining implementation, where we're currently carting around both block (future) objects and their corresponding metadata, and we're trying to surface metadata `Block` APIs for an `ObjectRef[Block]` via a separate `BlockMetadata`, which is clunky.

[Eric]
> Yes, I considered this, but it is less explicit. The current implementation is extremely clear about what is going on, which is why I chose it.

We can talk about this more later, but we might want to address this before we invest a lot into the current model. IMO the block metadata conceptually belongs to the block, even when just looking at the Block API, which exposes a bunch of metadata accessors. Splitting the block metadata out from the block abstraction results in a lot of duplicated logic: whenever we're getting or manipulating a block or list of blocks, we have to do the same to the sidecar metadata or list of metadatas, and we're always having to zip lists of blocks and their corresponding metadata together, or return a tuple containing blocks and their corresponding metadata. That's all a colocation code smell that suggests that those two abstractions should be consolidated. I actually think that this would make the implementation more clear, and I think that it would eliminate a lot of duplicated logic and bookkeeping, and will make future development simpler.

I can take a look at a time-boxed refactor later this week if that sounds good to you.

File: python/ray/experimental/data/impl/block_list.py
Comment (ericl) [reply python/ray/experimental/data/impl/block_list.py:9]: It's needed separately since it has to be accessed before the block has been computed. But I can move it.

File: python/ray/experimental/data/impl/compute.py
Comment (ericl) [reply python/ray/experimental/data/impl/compute.py:41]: I don't think this makes it more readable.

File: python/ray/experimental/data/impl/shuffle.py
Comment (ericl) [reply python/ray/experimental/data/impl/shuffle.py:64]: Eh turns out this breaks since you have to cast it to a list for Ray anyways, so not doing it.

**Comment by ericl** [comment]
> Splitting the block metadata out from the block abstraction results in a lot of duplicated logic

I think you're missing the main point here, in that metadata has separate lifetime from the block. In particular, it has to be accessible *before* the block is even generated. Hence, coupling them in the same object is introducing confusion about the lifetime.

I'm also not optimizing for code duplication here, but clarity and explicit behavior. I don't think the few extra for loops is worth the tradeoff.

File: python/ray/experimental/data/impl/shuffle.py
Comment (ericl) [reply python/ray/experimental/data/impl/shuffle.py:64]: Fixed

**Comment by clarkzinzow** [comment]
@ericl I definitely don't want to block this PR on this, and will review the PR as is with the current data model, but to address your points...

>I think you're missing the main point here, in that metadata has separate lifetime from the block. In particular, it has to be accessible before the block is even generated. Hence, coupling them in the same object is introducing confusion about the lifetime.

I'm not missing that point, I understand that the block metadata and data have different lifetimes. What I'm suggesting is that you take the `BlockList` abstraction's `blocks` and `metadata` structure, which is already coupling different lifetimes in the same object, and lower that down to the block level:
```python
class Block[T]:
    data: ObjectRef[BlockData[T]]
    metadata: BlockMetadata
```
where `BlockData[T]` is the current `Block[T]` API without the metadata accessors. The compute layer would still split these `data` and `metadata` fields out in order to flatten the Ray futures out into task arguments, but I'm contending that this data model will be nicer to work with in the layer above the compute layer and when adding to the data model on top of `Block`. 

>I'm also not optimizing for code duplication here, but clarity and explicit behavior. I don't think the few extra for loops is worth the tradeoff.

I'd agree with you wholeheartedly if I thought that was the actual tradeoff. I'm only trying to avoid future implementation complexity and subtle bugs that could be eliminated with a data model tweak early in the project, since I've seen (and implemented) both of these models in the past, and have found a tight coupling of the data future and the metadata to be better.

Anyways, I'm fine with going ahead with the current data model for now, we can reexamine it if it proves to be cumbersome to work with.

**Comment by ericl** [comment]
> Anyways, I'm fine with going ahead with the current data model for now, we can reexamine it if it proves to be cumbersome to work with.

I can see where you're coming from (and I've thought of those same advantages), but there are complications, in some cases you can also have BlockMetadata by itself where the Block computation isn't triggered yet (as in LazyBlockList). The call I'm making is that managing a collection of blocks with an abstraction like LazyBlockList is probably better than a plain list of more featureful Block classes. This is also looking forwards at supporting more advanced pipelining / streaming of transformations.

**Review by scv119 — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]
Rubber stamping for now with my WIP review, will posthoc review after merged. 😄

**Comment by ericl** [comment]
You can leave follow-up comments here: https://github.com/ray-project/ray/pull/16826

**Pull request merged**
